wrappers/clang-scan-deps-wrapper.sh

```diff
@@ -0,0 +1,71 @@
+#!/bin/sh
+#
+# Copyright (c) 2024 Martin Storsjo
+#
+# Permission to use, copy, modify, and/or distribute this software for any
+# purpose with or without fee is hereby granted, provided that the above
+# copyright notice and this permission notice appear in all copies.
+#
+# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+get_dir() {
+    target="$1"
+    while [ -L "$target" ]; do
+        cd "$(dirname "$target")"
+        target="$(readlink "$(basename "$target")")"
+    done
+    cd "$(dirname "$target")"
+    pwd
+}
+
+DIR="$(get_dir "$0")"
+BASENAME="$(basename "$0")"
+TARGET="${BASENAME%-clang-scan-deps}"
+DEFAULT_TARGET=x86_64-w64-mingw32
+if [ "$TARGET" = "$BASENAME" ]; then
+    TARGET=$DEFAULT_TARGET
+fi
```

Comment:
I think `clang-scan-deps` itself is target irrelevant (like other `llvm-*` tools), the arguments after `--` must be the actual command line to generate object files. In case we are using unprefixed compiler and  `TARGET` here is used for deps scanning, we may get wrong results if the actual default target of the compiler is not `TARGET`.
So I think if we are using unprefixed compiler, we should not add `-target` for deps scanning.

Reply:
My intent here, is that if `clang-scan-deps-wrapper.sh` is invoked, it means that you want whatever behaviours that the wrapper implies - i.e. defaulting to a mingw target.

On unix, if you'd invoke e.g. `x86_64-w64-mingw32-clang-scan-deps -- clang ...`, we would invoke `clang-scan-deps-real -- clang -target x86_64-w64-mingw32 ...`, which I think is a reasonable result. If the user did provide a custom `-target` option on the command line, it will come after the `-target` we injected, so the user provided option should take effect.

If you invoke `x86_64-w64-mingw32-clang-scan-deps -- armv7-w64-mingw32-clang ...`, then we end up invoking `clang-scan-deps-real -- armv7-w64-mingw32-clang -target armv7-w64-mingw32`, which probably is the most accurate interpretation of that situation.

One specific corner case regarding defaults here, is that I've tried to design the wrapper setup, so that you could use any clang executables with any defaults, e.g. the official releases that default to targeting MSVC. If you, on Windows, invoke `clang-scan-deps -- clang ...`, then `clang-scan-deps` is a wrapper too, which will default to injecting `-target <arch>-w64-mingw32`. This makes sure it works even if `clang-scan-deps-real.exe` is a build that defaults to MSVC targets.

On Windows, `clang-scan-deps.exe` is an executable wrapper, and the real clang binary is `clang-scan-deps-real.exe`. On Unix, we leave `clang-scan-deps` as it was, as the original vanilla binary, so the case of default implicit target shouldn't really matter.

Reply:
> On unix, if you'd invoke e.g. `x86_64-w64-mingw32-clang-scan-deps -- clang ...`, we would invoke `clang-scan-deps-real -- clang -target x86_64-w64-mingw32 ...`

We can see how cmake invoking `clang-scan-deps` here:
https://github.com/Kitware/CMake/blob/v3.29.0/Modules/Compiler/Clang-CXX.cmake#L49-L56
The arguments after `--` is almost same as how cmake produce the object file:
https://github.com/Kitware/CMake/blob/v3.29.0/Modules/CMakeCXXInformation.cmake#L284

So basically, we first invoke `x86_64-w64-mingw32-clang-scan-deps -- clang ...` to get deps to know the compiling order of source files, then we invoke exactly same the `clang ...` command the produce object files.
Now the problem is we scan deps targeting `x86_64-w64-mingw32`, but compile the sources targeting default (on unix, it is `x86_64-unknown-linux-gnu`).

Reply:
> Now the problem is we scan deps targeting x86_64-w64-mingw32, but compile the sources targeting default (on unix, it is x86_64-unknown-linux-gnu).

Yes, that's true.

But normally this wouldn't happen on its own - as CMake chooses to invoke `<triple>-clang-scan-deps` because the compiler is called `<triple>-clang` or similar, right?

Now if we put it another way - if you invoke `<triple>-<tool>`, that generally also expects getting `<triple>` flavoured defaults within that tool. So it's not entirely farfetched, to expect that if you run `<triple>-clang-scan-deps`, that you'd want `<triple>` as some sort of default.

Now in this particular case, on Unix, you're right that this wouldn't match what the actual compilation command would do. But then again, why would the caller (the user, or cmake) go out of their way to call `clang-scan-deps` in that form?

Anyway, when the command following the `--` is a plain `clang`, you're right that it really does do compilation for a Unix target, so injecting our own target isn't right. But then we also shouldn't be injecting `-stdlib=libc++` either, since that's also specific for the mingw targets in the llvm-mingw toolchain.

Secondly, what if you're invoking `x86_64-w64-mingw32-clang-scan-deps -- nonobvious-wrapper-script-clang ...`. We don't have such cases in llvm-mingw right now, but it's not implausible to have wrappers that aren't a plain triple prefix. In this case, the command `nonobvious-wrapper-script-clang` does invoke clang with some non-default settings, and we don't really know what they are. In that case, my point is that we're invoking `x86_64-w64-mingw32-clang-scan-deps` because we want to apply mingw style defaults, even if that's not obvious from the command after `--`.

In any case, what kind of behaviour to imply in this case is a bit of theoretical corner case - I think both behaviours can be considered reasonable. I can try to implement what you're suggesting, that we entirely ignore the prefix of the `clang-scan-deps` wrapper and only inspect the nested command.

But the other case I mentioned, where we do need to force our defaults, is in Windows builds. Because as a design feature, `clang-18.exe` and `clang-scan-deps-real.exe` can be binaries that have any defaults (they could default to a MSVC target). So when we invoke `clang-scan-deps -- clang ...`, where `clang-scan-deps` is our wrapper, we do want to force adding our default `-target`.

Reply:
Ok, see a29cee57f0e8dd3397a96e948c3ffa4e9b18b1cc for an attempt at doing what you suggest here.

I guess this could be fine with me, too. (And it does simplify the code, so that's always nice of course.)

But as for intended behaviour, I have another hypothetical example to consider, as a thought experiment:

If we're run as `x86_64-w64-mingw32-clang-scan-deps -- clang ...`, we'll invoke `clang-scan-deps -- clang ...` with no extra `-target` or any extra arguments. If we're run as `x86_64-w64-mingw32-clang-scan-deps -- armv7-w64-mingw32-clang ...`, we'll invoke `clang-scan-deps -- armv7-w64-mingw32-clang -target armv7-w64-mingw32 -stdlib=libc++ ...`. Those are the simple cases.

But what if we're invoked as `x86_64-w64-mingw32-clang-scan-deps -- aarch64-linux-gnu-clang ...`, we'll invoke `clang-scan-deps -- aarch64-linux-gnu-clang -target aarch64-linux-gnu -stdlib=libc++ ...`. But the `-stdlib=libc++` option doesn't necessarily apply here. So if we continue this thought experiment, we should maybe inspect the target triple and only apply our default `-stdlib=libc++` if the target looks like a mingw target triple.

It sounds a bit far fetched - but my point is that once we're invoked via the mingw specific wrapper, it's not entirely unreasonable to assume mingw specific target behaviours anyway.

Reply:
> my point is that once we're invoked via the mingw specific wrapper, it's not entirely unreasonable to assume mingw specific target behaviours anyway.

I fully support this. My main argument here is that `clang-scan-deps` is actually target independent, so the prefix should not have side effect, it is just a conventional naming scheme to avoid conflicit with host `clang-scan-deps`. 

I also support that the `clang-scan-deps` wrapper should have some defaults so it is consistent with `clang`, and is consistent within the whole toolchain. Like you pointed out specifically the windows case.

> Because as a design feature, `clang-18.exe` and `clang-scan-deps-real.exe` can be binaries that have any defaults (they could default to a MSVC target). So when we invoke `clang-scan-deps -- clang ...`, where `clang-scan-deps` is our wrapper, we do want to force adding our default `-target`.

Out of topic: can wrappers provided by llvm-mingw be used with LLVM official build, to support mingw target build?

Reply:
> I fully support this. My main argument here is that clang-scan-deps is actually target independent, so the prefix should not have side effect, it is just a conventional naming scheme to avoid conflicit with host clang-scan-deps.

Ok, fair enough.

Anyway, my follow-up is mostly that once we remove the case where we always default to setting a mingw target by default, the `-stdlib=libc++` flag stood out like a weird extra addition there. I tested another addition to it, to make that conditional like this: https://github.com/mstorsjo/llvm-mingw/commit/3cc7cda9097e1a3b7c0b2e3aaaf77c94d32242ca
I'm not entirely sure which way I prefer here - I presume you prefer the state after https://github.com/mstorsjo/llvm-mingw/commit/a29cee57f0e8dd3397a96e948c3ffa4e9b18b1cc at least, with or without that change?

> Out of topic: can wrappers provided by llvm-mingw be used with LLVM official build, to support mingw target build?

In theory, yes, although I haven't actually tried to set it up.

On Unix, you can nowadays build a llvm-mingw toolchain around a preexisting install of Clang, see https://github.com/mstorsjo/llvm-mingw/blob/master/Dockerfile.system-clang for an example of this, running `build-all.sh` with `--host-clang`.

For toolchains running on Windows, there's no such premade setup script though, and it's a little tricky as we mostly cross compile those toolchains from Unix. But I guess it should be possible to alter that, like the `build-all.sh` case with `--host-clang` above.

But e.g. for a preexisting llvm-mingw toolchain on Windows, it should be possible to replace any of the actual Clang/LLVM binaries with ones from any build. E.g. for the main clang executable, in our windows toolchains, these are named e.g. `clang-18.exe`, and this name is hardcoded in the executable wrappers. Swapping that out for any other Clang 18.x build should be a complete drop-in. If you want to use a Clang 19.x build, it's fine to just drop it in and rename it to `clang-18.exe`, that should work, even if it's misnamed. But if you do that, you'll need to rename or make a copy of `<prefix>/lib/clang/18` and name it `<prefix>/lib/clang/19` for the new Clang to find it.

Reply:
> I'm not entirely sure which way I prefer here - I presume you prefer the state after https://github.com/mstorsjo/llvm-mingw/commit/a29cee57f0e8dd3397a96e948c3ffa4e9b18b1cc at least, with or without that change?

Yes, both look good to me. Either one will work correctly since we expect user (cmake) always use prefixed tools consistently.